src/v/storage/mvlog/readable_segment.h

```diff
@@ -22,14 +25,19 @@ class readable_segment {
     explicit readable_segment(file* f)
       : file_(f) {}
 
-    std::unique_ptr<segment_reader> make_reader();
+    std::unique_ptr<segment_reader> make_reader(const version_id tid);
     size_t num_readers() const { return num_readers_; }
+    const versioned_gap_list& gaps() const { return gaps_; }
+    versioned_gap_list* mutable_gaps() { return &gaps_; }
```

Comment:
nice to have the mutable prefix. not sure yet how this is used but keeping in the spirit of mvcc/copy-on-write i wonder if we can arrange for gap modification to merely create a new object that has the modification in it?

Reply:
I liked version ids because they allowed to get a "consistent" view of the log.

Here is a scenario where it would be useful:

1. Fetch request arrives to redpanda
2. We validate min/max offset, leader epoch, all good, we forward the request to the storage layer to read offset [5, 10]
3. Concurrently: Log is being truncated, new entries are being appended
4. What data will the fetch request get back? Can you reason about the correctness?

If at step 2 we would have captured a log versioned id (true mvcc), then you wouldn't have to think about potential problems introduced by step 3. It would also make it much easier to reason about correctness.

I believe this problem exists today in Redpanda. We don't lock the log before the step 2, do we?

Reply:
Likely this wasn't part of original requirements but it might be good time to think about it. PS. build failures.